Rust Dockerfiles parser fails on base image with hash and build arg

Thanks for 2.27.0 release. Unfortunately I was late in testing the RC.

**Describe the bug**
Using the rust Dockerfile parser, this Dockerfile fails to build:

    ARG DOCKER_IO_MIRROR=docker.io
    FROM $DOCKER_IO_MIRROR/dperson/samba@sha256:e1d2a7366690749a7be06f72bdbf6a5a7d15726fc84e4e4f41e967214516edfd

The same build works fine using the python parser.

**Pants version**
2.27.0

**OS**
Linux

**Additional info**

I have added some tests to expose the issue here: https://github.com/bdabelow/pants/commit/8311df26eec4ca4d23909f4425e65dfeb8d88561

Minimal repro is here: https://github.com/bdabelow/pants-repro/tree/main/rust-dockerfile-parser

Log:

        > pants package ::
        12:16:12.97 [INFO] Initializing scheduler...
        12:16:14.87 [INFO] Scheduler initialized.
        12:16:14.97 [INFO] Completed: Scheduling: Test binary /usr/bin/docker.
        12:16:14.99 [INFO] Completed: Scheduling: Test binary /bin/docker.
        12:16:15.01 [ERROR] 1 Exception encountered:
        
        Engine traceback:
          in root
            ..
          in pants.core.goals.package.package_asset
            `package` goal
        
        Traceback (most recent call last):
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/core/goals/package.py", line 195, in package_asset
            packages = await MultiGet(
                       ^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/engine/internals/selectors.py", line 356, in MultiGet
            return await _MultiGet(tuple(__arg0))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/engine/internals/selectors.py", line 163, in __await__
            result = yield self.gets
                     ^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/core/goals/package.py", line 146, in environment_aware_package
            package = await Get(
                      ^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/engine/internals/selectors.py", line 113, in __await__
            result = yield self
                     ^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/backend/docker/goals/package_image.py", line 394, in build_docker_image
            context, wrapped_target = await MultiGet(
                                      ^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/engine/internals/selectors.py", line 386, in MultiGet
            return await _MultiGet((__arg0, __arg1))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/engine/internals/selectors.py", line 163, in __await__
            result = yield self.gets
                     ^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/backend/docker/util_rules/docker_build_context.py", line 384, in create_docker_build_context
            return DockerBuildContext.create(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/backend/docker/util_rules/docker_build_context.py", line 137, in create
            stage_names, tags_values = cls._get_stages_and_tags(
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/home/axccu/.cache/nce/6d61748cea187199dc28418157b2121ccbbb44e4af2ee6fddf8c04bbed73f76d/bindings/venvs/2.27.0/lib/python3.11/site-packages/pants/backend/docker/util_rules/docker_build_context.py", line 186, in _get_stages_and_tags
            raise DockerBuildContextError(
        pants.backend.docker.util_rules.docker_build_context.DockerBuildContextError: Failed to parse Dockerfile baseimage tag for stage stage0 in // target, from image ref: docker.io.

--- a/docs/notes/2.33.x.md
+++ b/docs/notes/2.33.x.md
@@ -34,6 +34,8 @@ Fixed broken troubleshooting documentation links.
 
 The `docker_image` target now supports capturing file and directory artifacts from Docker builds using the `output_files` and `output_directories` fields. This supports workflows where a Dockerfile stage creates a build artifact(s) that should be consumed by other Pants targets, using the same behavior as the `shell_command` or `adhoc_tool` targets. This feature uses the BuildKit local output exporter and requires `[docker].use_buildx = true`.
 
+Fixed the Rust Dockerfile parser treating a build arg used as only part of a `FROM` image reference (such as the registry prefix) as the base image tag, which caused specs like `FROM $REGISTRY/image@sha256:...` to fail ([#22443](https://github.com/pantsbuild/pants/issues/22443)).
+
 #### NEW: OpenTelemetry
 
 Add a new `pants.backend.observability.opentelemetry` backend to report work unit tracing to OpenTelemetry.

--- a/src/python/pants/backend/docker/subsystems/dockerfile_parser_test.py
+++ b/src/python/pants/backend/docker/subsystems/dockerfile_parser_test.py
@@ -214,6 +214,11 @@ def test_baseimage_tags(rule_runner: RuleRunner) -> None:
                 "v0.54.0@sha256:d1f0463b35135852308ea815c2ae54c1734b876d90288ce35828aeeff9899f9d\n"
                 "FROM $PYTHON_VERSION AS python\n"
                 "FROM python:$VERSION\n"
+                "FROM $MIRROR/dperson/samba@sha256:"
+                "e1d2a7366690749a7be06f72bdbf6a5a7d15726fc84e4e4f41e967214516edfd\n"
+                "FROM $REGISTRY/$IMAGE\n"
+                "FROM registry/$IMAGE\n"
+                "FROM $IMAGE@sha256:d1f0463b35135852308ea815c2ae54c1734b876d90288ce35828aeeff9899f9d\n"
             ),
         }
     )
@@ -226,6 +231,10 @@ def test_baseimage_tags(rule_runner: RuleRunner) -> None:
         "stage3 v0.54.0",
         "python build-arg:PYTHON_VERSION",  # Parse tag from build arg.
         "stage5 $VERSION",
+        "stage6",
+        "stage7 build-arg:IMAGE",
+        "stage8 build-arg:IMAGE",
+        "stage9",
     )
 
 

--- a/src/rust/dep_inference/src/dockerfile/from.rs
+++ b/src/rust/dep_inference/src/dockerfile/from.rs
@@ -7,7 +7,8 @@ use crate::dockerfile::{ChildBehavior, Visitor};
 
 pub(crate) struct TagCollector<'a> {
     pub tag: Tag<'a>,
-    is_arg: bool,
+    pub image_arg: Option<&'a str>,
+    image_name: Option<&'a str>,
     code: &'a str,
 }
 
@@ -15,7 +16,8 @@ impl<'a> TagCollector<'a> {
     pub fn new(code: &'a str) -> Self {
         Self {
             tag: Tag::None,
-            is_arg: false,
+            image_arg: None,
+            image_name: None,
             code,
         }
     }
@@ -31,20 +33,27 @@ pub(crate) enum Tag<'a> {
 
 impl Visitor for TagCollector<'_> {
     fn visit_variable(&mut self, node: Node) -> ChildBehavior {
-        if self.is_arg {
-            self.tag = Tag::BuildArg(code::at(self.code, node));
+        if let Some(image_name) = self.image_name {
+            let var = code::at(self.code, node);
+            let is_var = |s: &str| {
+                s.strip_prefix('$') == Some(var)
+                    || s.strip_prefix("${").and_then(|b| b.strip_suffix('}')) == Some(var)
+            };
+            // Whole image name is a build arg: an inferred dependency.
+            if is_var(image_name) {
+                self.image_arg = Some(var);
+            }
+            // Final path segment is a build arg: the base image tag.
+            let last_segment = image_name.rsplit('/').next().unwrap_or(image_name);
+            if is_var(last_segment) {
+                self.tag = Tag::BuildArg(var);
+            }
         }
         ChildBehavior::Ignore
     }
 
-    fn visit_image_spec(&mut self, node: Node) -> ChildBehavior {
-        if code::at(self.code, node).trim().starts_with('$') {
-            self.is_arg = true;
-        }
-        ChildBehavior::Visit
-    }
-
-    fn visit_image_name(&mut self, _: Node) -> ChildBehavior {
+    fn visit_image_name(&mut self, node: Node) -> ChildBehavior {
+        self.image_name = Some(code::at(self.code, node));
         if self.tag == Tag::None {
             self.tag = Tag::Default
         }
@@ -52,6 +61,7 @@ impl Visitor for TagCollector<'_> {
     }
 
     fn visit_image_tag(&mut self, node: Node) -> ChildBehavior {
+        self.image_arg = None;
         self.tag = code::at(self.code, node)
             .strip_prefix(':')
             .map(Tag::Explicit)
@@ -60,7 +70,8 @@ impl Visitor for TagCollector<'_> {
     }
 
     fn visit_image_digest(&mut self, _node: Node) -> ChildBehavior {
-        if self.tag == Tag::Default {
+        self.image_arg = None;
+        if matches!(self.tag, Tag::Default | Tag::BuildArg(_)) {
             self.tag = Tag::None
         }
         ChildBehavior::Ignore

--- a/src/rust/dep_inference/src/dockerfile/mod.rs
+++ b/src/rust/dep_inference/src/dockerfile/mod.rs
@@ -131,15 +131,15 @@ impl Visitor for DockerFileInfoCollector<'_> {
         let mut cursor = node.walk();
         let mut tag_collector = TagCollector::new(self.code);
         tag_collector.walk(&mut cursor);
+        if let Some(arg) = tag_collector.image_arg {
+            self.from_image_build_args.extend(
+                self.get_build_arg_reference(arg)
+                    .map(|val| format!("{arg}={val}")),
+            );
+        }
         let tag = match tag_collector.tag {
             Tag::Default => Some("latest".to_string()),
-            Tag::BuildArg(arg) => {
-                self.from_image_build_args.extend(
-                    self.get_build_arg_reference(arg)
-                        .map(|val| format!("{arg}={val}")),
-                );
-                Some(format!("build-arg:{arg}"))
-            }
+            Tag::BuildArg(arg) => Some(format!("build-arg:{arg}")),
             Tag::Explicit(e) => Some(e.to_string()),
             Tag::None => None,
         };

--- a/src/rust/dep_inference/src/dockerfile/tests.rs
+++ b/src/rust/dep_inference/src/dockerfile/tests.rs
@@ -78,6 +78,26 @@ fn from_instructions() {
         "FROM gcr.io/tekton-releases/github.com/tektoncd/operator/cmd/kubernetes/operator:v0.54.0@sha256:d1f0463b35135852308ea815c2ae54c1734b876d90288ce35828aeeff9899f9d",
         [("stage0", Some("v0.54.0"))],
     );
+    assert_from_tags(
+        "ARG MIRROR=docker.io\nFROM $MIRROR/dperson/samba@sha256:e1d2a7366690749a7be06f72bdbf6a5a7d15726fc84e4e4f41e967214516edfd",
+        [("stage0", None)],
+    );
+    assert_from_tags(
+        "ARG MIRROR=docker.io\nFROM ${MIRROR}/python:3.10",
+        [("stage0", Some("3.10"))],
+    );
+    assert_from_tags(
+        "FROM $REGISTRY/$IMAGE",
+        [("stage0", Some("build-arg:IMAGE"))],
+    );
+    assert_from_tags(
+        "FROM registry/$IMAGE",
+        [("stage0", Some("build-arg:IMAGE"))],
+    );
+    assert_from_tags(
+        "FROM $IMAGE@sha256:d1f0463b35135852308ea815c2ae54c1734b876d90288ce35828aeeff9899f9d",
+        [("stage0", None)],
+    );
 }
 
 #[test]
@@ -157,4 +177,6 @@ FROM $DECOY
             "BASE_IMAGE_5=should/allow/default-target-name",
         ],
     );
+    assert_from_image_build_args("ARG IMAGE=some/image\nFROM registry/$IMAGE", []);
+    assert_from_image_build_args("ARG IMAGE=some/image\nFROM $IMAGE:latest", []);
 }

